# Activity: Discover what is in your dataset

## Introduction

In this activity, you will discover characteristics of a dataset and use visualizations to analyze the data. This will develop and strengthen your skills in **exploratory data analysis (EDA)** and your knowledge of functions that allow you to explore and visualize data. 

EDA is an essential process in a data science workflow. As a data professional, you will need to conduct this process to better understand the data at hand and determine how it can be used to solve the problem you want to address. This activity will give you an opportunity to practice that process and prepare you for EDA in future projects.

In this activity, you are a member of an analytics team that provides insights to an investing firm. To help them decide which companies to invest in next, the firm wants insights into **unicorn companies**–companies that are valued at over one billion dollars. The data you will use for this task provides information on over 1,000 unicorn companies, including their industry, country, year founded, and select investors. You will use this information to gain insights into how and when companies reach this prestigious milestone and to make recommendations for next steps to the investing firm.

## Step 1: Imports

### Import libraries and packages 

First, import relevant Python libraries and modules. Use the `pandas` library and the `matplotlib.pyplot` module.

In [1]:
# Import libraries and packages

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

### Load the dataset into a DataFrame

The dataset provided is in the form of a csv file named `Unicorn_Companies.csv` and contains a subset of data on unicorn companies. As shown in this cell, the dataset has been automatically loaded in for you. You do not need to download the .csv file, or provide more code, in order to access the dataset and proceed with this lab. Please continue with this activity by completing the following instructions.

In [2]:
# RUN THIS CELL TO IMPORT YOUR DATA.

### YOUR CODE HERE ###
companies = pd.read_csv("resources_repo/Modified_Unicorn_Companies.csv")

<details>
  <summary><h4><strong>Hint 1</strong></h4></summary>

Refer to what you learned about [loading data](https://www.coursera.org/learn/go-beyond-the-numbers-translate-data-into-insight/supplement/MdTG2/reference-guide-import-datasets-using-python) in Python.

</details>

<details>
  <summary><h4><strong>Hint 2</strong></h4></summary>

Use the function in the `pandas` library that allows you to read data from a csv file and load the data into a DataFrame.


</details>

<details>
  <summary><h4><strong>Hint 3</strong></h4></summary>

Use the `read_csv()` function from the `pandas` library. 

</details>

## Step 2: Data exploration

### Display the first 10 rows of the data

Next, explore the dataset and answer questions to guide your exploration and analysis of the data. To begin, display the first 10 rows of the data to get an understanding of how the dataset is structured.

In [3]:
# Display the first 10 rows of the data

companies.head(10)

Company  Valuation Date Joined                             Industry  \
0     Bytedance        180  2017-04-07              Artificial intelligence   
1        SpaceX        100  2012-12-01                                Other   
2         SHEIN        100  2018-07-03      E-commerce & direct-to-consumer   
3        Stripe         95  2014-01-23                              FinTech   
4        Klarna         46  2011-12-12                              Fintech   
5         Canva         40  2018-01-08         Internet software & services   
6  Checkout.com         40  2019-05-02                              Fintech   
7     Instacart         39  2014-12-30  Supply chain, logistics, & delivery   
8     JUUL Labs         38  2017-12-20                    Consumer & retail   
9    Databricks         38  2019-02-05        Data management and analytics   

            City  Country/Region      Continent  Year Founded Funding  \
0        Beijing           China           Asia          2012     $8B   
1      Hawthorne   United States  North America          2002     $7B   
2       Shenzhen           China           Asia          2008     $2B   
3  San Francisco   United States  North America          2010     $2B   
4      Stockholm          Sweden         Europe          2005     $4B   
5    Surry Hills       Australia        Oceania          2012   $572M   
6         London  United Kingdom         Europe          2012     $2B   
7  San Francisco   United States  North America          2012     $3B   
8  San Francisco   United States  North America          2015    $14B   
9  San Francisco   United States  North America          2013     $3B   

                                    Select Investors  
0  Sequoia Capital China, SIG Asia Investments, S...  
1  Founders Fund, Draper Fisher Jurvetson, Rothen...  
2  Tiger Global Management, Sequoia Capital China...  
3        Khosla Ventures, LowercaseCapital, capitalG  
4  Institutional Venture Partners, Sequoia Capita...  
5  Sequoia Capital China, Blackbird Ventures, Mat...  
6  Tiger Global Management, Insight Partners, DST...  
7  Khosla Ventures, Kleiner Perkins Caufield & By...  
8                            Tiger Global Management  
9  Andreessen Horowitz, New Enterprise Associates...

<details>
  <summary><h4><strong>Hint 1</strong></h4></summary>

Refer to [the content about exploratory data analysis in Python](https://www.coursera.org/learn/go-beyond-the-numbers-translate-data-into-insight/lecture/kfl9b/find-stories-using-the-six-exploratory-data-analysis-practices).

</details>

<details>
  <summary><h4><strong>Hint 2</strong></h4></summary>

Use the function in the `pandas` library that allows you to get a specific number of rows from the top of a DataFrame.
 

</details>

<details>
  <summary><h4><strong>Hint 3</strong></h4></summary>

Use the `head()` function from the `pandas` library. 

</details>

**Question:** What do you think the "Date Joined" column represents?

Date Joined represents the date when the company became a unicorn company.

**Question:** What do you think the "Select Investors" column represents?

Select Investors represents the investors who have invested in the company.

### Assess the size of the dataset

Get a sense of how large the dataset is. The `size` property that DataFrames have can help.

In [5]:
# How large the dataset is

companies.size

10740

**Question:** What do you notice about the size of the dataset?

The dataset has 10740 as size attribute

### Determine the shape of the dataset

Identify the number of rows and columns in the dataset. The `shape` property that DataFrames have can help.

In [6]:
# Shape of the dataset

companies.shape

(1074, 10)

**Question:** What do you notice about the shape of the dataset?

The dataset has 1074 rows and 10 columns.

### Get basic information about the dataset

To further understand what the dataset entails, get basic information about the dataset, including the data type of values in each column. There is more than one way to approach this task. In this instance, use the `info()` function from `pandas`.

In [7]:
# Get information

companies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1074 entries, 0 to 1073
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Company           1074 non-null   object
 1   Valuation         1074 non-null   int64 
 2   Date Joined       1074 non-null   object
 3   Industry          1074 non-null   object
 4   City              1057 non-null   object
 5   Country/Region    1074 non-null   object
 6   Continent         1074 non-null   object
 7   Year Founded      1074 non-null   int64 
 8   Funding           1074 non-null   object
 9   Select Investors  1074 non-null   object
dtypes: int64(2), object(8)
memory usage: 84.0+ KB


**Question:** What do you notice about the type of data in the `Year Founded` column? Refer to the output from using `info()` above. Knowing the data type of this variable is helpful because it indicates what types of analysis can be done with that variable, how it can be aggregated with other variables, and so on.

The column `Year Founded` is of type `int64` a numeric data type. It should be a date type.

**Question:** What do you notice about the type of data in the `Date Joined` column? Refer to the output from using `info()` above. Knowing the data type of this variable is helpful because it indicates what types of analysis can be done with that variable and how the variable can be transformed to suit specific tasks.

The column `Date Joined` is of type `object` which is a string data type. It should be a date type.

## Step 3: Statistical tests

### Find descriptive statistics

Find descriptive statistics and structure your dataset. The `describe()` function from the `pandas` library can help. This function generates statistics for the numeric columns in a dataset. 

In [8]:
# Get descriptive statistics

companies.describe()


Valuation  Year Founded
count  1074.000000   1074.000000
mean      3.445996   2012.870577
std       8.544242      5.705494
min       1.000000   1919.000000
25%       1.000000   2011.000000
50%       2.000000   2014.000000
75%       3.000000   2016.000000
max     180.000000   2021.000000

**Question:** Based on the table of descriptive stats generated above, what do you notice about the minimum value in the `Year Founded` column? This is important to know because it helps you understand how early the entries in the data begin.

The minimum value in the `Year Founded` column is 1919. This is the earliest year in the dataset.

**Question:** What do you notice about the maximum value in the `Year Founded` column? This is important to know because it helps you understand the most recent year captured by the data. 

The maximum value in the `Year Founded` column is 2021. This is the most recent year in the dataset.

### Convert the `Date Joined` column to datetime

Use the `to_datetime()` function from the `pandas` library  to convert the `Date Joined` column to datetime. This splits each value into year, month, and date components. This is an important step in data cleaning, as it makes the data in this column easier to use in tasks you may encounter. To name a few examples, you may need to compare "date joined" between companies or determine how long it took a company to become a unicorn. Having "date joined" in datetime form would help you complete such tasks.

In [9]:
# Step 1: Use pd.to_datetime() to convert Date Joined column to datetime
# Step 2: Update the column with the converted values

companies['Date Joined'] = pd.to_datetime(companies['Date Joined'])


In [10]:
# Use .info() to confirm that the update actually took place

companies.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1074 entries, 0 to 1073
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Company           1074 non-null   object        
 1   Valuation         1074 non-null   int64         
 2   Date Joined       1074 non-null   datetime64[ns]
 3   Industry          1074 non-null   object        
 4   City              1057 non-null   object        
 5   Country/Region    1074 non-null   object        
 6   Continent         1074 non-null   object        
 7   Year Founded      1074 non-null   int64         
 8   Funding           1074 non-null   object        
 9   Select Investors  1074 non-null   object        
dtypes: datetime64[ns](1), int64(2), object(7)
memory usage: 84.0+ KB


### Create a `Year Joined` column

It is common to encounter situations where you will need to compare the year joined with the year founded. The `Date Joined` column does not just have year—it has the year, month, and date. Extract the year component from the `Date Joined` column and add those year components into a new column to keep track of each company's year joined.

In [11]:
# Step 1: Use .dt.year to extract year component from Date Joined column
# Step 2: Add the result as a new column named Year Joined to the DataFrame

companies['Year Joined'] = companies['Date Joined'].dt.year


In [12]:
# Use .head() to confirm that the new column did get added

companies.head()


Company  Valuation Date Joined                         Industry  \
0  Bytedance        180  2017-04-07          Artificial intelligence   
1     SpaceX        100  2012-12-01                            Other   
2      SHEIN        100  2018-07-03  E-commerce & direct-to-consumer   
3     Stripe         95  2014-01-23                          FinTech   
4     Klarna         46  2011-12-12                          Fintech   

            City Country/Region      Continent  Year Founded Funding  \
0        Beijing          China           Asia          2012     $8B   
1      Hawthorne  United States  North America          2002     $7B   
2       Shenzhen          China           Asia          2008     $2B   
3  San Francisco  United States  North America          2010     $2B   
4      Stockholm         Sweden         Europe          2005     $4B   

                                    Select Investors  Year Joined  
0  Sequoia Capital China, SIG Asia Investments, S...         2017  
1  Founders Fund, Draper Fisher Jurvetson, Rothen...         2012  
2  Tiger Global Management, Sequoia Capital China...         2018  
3        Khosla Ventures, LowercaseCapital, capitalG         2014  
4  Institutional Venture Partners, Sequoia Capita...         2011

## Step 4: Results and evaluation


### Take a sample of the data

It is not necessary to take a sample of the data in order to conduct the visualizations and EDA that follow. But you may encounter scenarios in the future where you will need to take a sample of the data due to time and resource limitations. For the purpose of developing your skills around sampling, take a sample of the data and work with that sample for the next steps of analysis you want to conduct. Use the `sample()` function for this task.

- Use `sample()` with the `n` parameter set to `50` to randomly sample 50 unicorn companies from the data. Be sure to specify the `random_state` parameter to ensure reproducibility of your work. Save the result to a variable called `companies_sampled`.

In [13]:
# Sample the data

companies_sampled = companies.sample(50, random_state=1)
companies_sampled.head()


Company  Valuation Date Joined                         Industry  \
635  Rad Power Bikes          1  2021-10-28            Auto & transportation   
428           Zenoti          2  2020-12-15     Internet software & services   
49      ServiceTitan         10  2018-11-14     Internet software & services   
204          Dutchie          4  2021-03-16  E-commerce & direct-to-consumer   
998            PLACE          1  2021-11-17     Internet software & services   

           City Country/Region      Continent  Year Founded Funding  \
635     Seattle  United States  North America          2007   $329M   
428    Bellevue  United States  North America          2010   $331M   
49     Glendale  United States  North America          2012     $1B   
204        Bend  United States  North America          2017   $603M   
998  Bellingham  United States  North America          2020   $100M   

                                      Select Investors  Year Joined  
635  Durable Capital Partners, Cercano Management, ...         2021  
428  Norwest Venture Partners, Accel, Tiger Global ...         2020  
49   Bessemer Venture Partners, ICONIQ Capital, Bat...         2018  
204  Casa Verde Capital, Gron Ventures, Thrity Five...         2021  
998                 Goldman Sachs Asset Management, 3L         2021

In [19]:
companies_sampled.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 50 entries, 635 to 133
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Company           50 non-null     object        
 1   Valuation         50 non-null     int64         
 2   Date Joined       50 non-null     datetime64[ns]
 3   Industry          50 non-null     object        
 4   City              49 non-null     object        
 5   Country/Region    50 non-null     object        
 6   Continent         50 non-null     object        
 7   Year Founded      50 non-null     int64         
 8   Funding           50 non-null     object        
 9   Select Investors  50 non-null     object        
 10  Year Joined       50 non-null     int64         
 11  Years to Unicorn  50 non-null     int64         
dtypes: datetime64[ns](1), int64(4), object(7)
memory usage: 5.1+ KB


In [14]:
# number of companies in companies_sampled
companies_sampled.shape

(50, 11)

### Visualize the time it took companies to reach unicorn status

Visualize the longest time it took companies to reach unicorn status for each industry represented in the sample. To create a bar plot to visualize this, use the `bar()` function from the `matplotlib.pyplot` module. You'll first need to prepare the data.

In [25]:
# Prepare data for plot
# covert year founded to datetime
companies_sampled['Year Founded'] = pd.to_datetime(companies_sampled['Year Founded'])
# extract year from year founded
companies_sampled['Year Founded'] = companies_sampled['Year Founded'].dt.year
companies_sampled.info()


<class 'pandas.core.frame.DataFrame'>
Int64Index: 50 entries, 635 to 133
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Company           50 non-null     object        
 1   Valuation         50 non-null     int64         
 2   Date Joined       50 non-null     datetime64[ns]
 3   Industry          50 non-null     object        
 4   City              49 non-null     object        
 5   Country/Region    50 non-null     object        
 6   Continent         50 non-null     object        
 7   Year Founded      50 non-null     int64         
 8   Funding           50 non-null     object        
 9   Select Investors  50 non-null     object        
 10  Year Joined       50 non-null     int64         
 11  Years to Unicorn  50 non-null     int64         
dtypes: datetime64[ns](1), int64(4), object(7)
memory usage: 5.1+ KB


In [26]:
# add column for years to get to Unicorn status
companies_sampled['Years to Unicorn'] = companies_sampled['Year Joined'] - companies_sampled['Year Founded']
companies_sampled.head()

Company  Valuation Date Joined                         Industry  \
635  Rad Power Bikes          1  2021-10-28            Auto & transportation   
428           Zenoti          2  2020-12-15     Internet software & services   
49      ServiceTitan         10  2018-11-14     Internet software & services   
204          Dutchie          4  2021-03-16  E-commerce & direct-to-consumer   
998            PLACE          1  2021-11-17     Internet software & services   

           City Country/Region      Continent  Year Founded Funding  \
635     Seattle  United States  North America          1970   $329M   
428    Bellevue  United States  North America          1970   $331M   
49     Glendale  United States  North America          1970     $1B   
204        Bend  United States  North America          1970   $603M   
998  Bellingham  United States  North America          1970   $100M   

                                      Select Investors  Year Joined  \
635  Durable Capital Partners, Cercano Management, ...         2021   
428  Norwest Venture Partners, Accel, Tiger Global ...         2020   
49   Bessemer Venture Partners, ICONIQ Capital, Bat...         2018   
204  Casa Verde Capital, Gron Ventures, Thrity Five...         2021   
998                 Goldman Sachs Asset Management, 3L         2021   

     Years to Unicorn  
635                51  
428                50  
49                 48  
204                51  
998                51

In [29]:
# group the companies by industry to get the maximum years to unicorn
max_years_to_unicorn = companies_sampled.groupby('Industry')['Years to Unicorn'].max().sort_values(ascending=False)

In [30]:
# Create bar plot
# with the various industries as the categories of the bars
# and the time it took to reach unicorn status as the height of the bars

plt.bar(x=companies_sampled['industry'], height=companies_sampled['max_years_to_unicorn'])


# Set title

plt.title('Years to Unicorn by Industry')


# Set x-axis label

plt.xlabel('Industry')


# Set y-axis label

plt.ylabel('Years to Unicorn')


# Rotate labels on the x-axis as a way to avoid overlap in the positions of the text

plt.xticks(rotation=90)


# Display the plot

plt.show()


KeyError: 'industry'

<details>
  <summary><h4><strong>Hint 1</strong></h4></summary>

To prepare the data for modeling, begin by creating a column that represents the number of years it took each company to reach unicorn status. 

</details>

<details>
  <summary><h4><strong>Hint 2</strong></h4></summary>
    
To prepare the data for modeling, group the dataframe by industry and get the maximum value in the newly created column for each industry.

</details>

<details>
  <summary><h4><strong>Hint 3</strong></h4></summary>

You can use the `plt.bar()` to create the bar plot, passing in the categories and heights of the bars.

You can use `plt.title()`, `plt.xlabel()`, and `plt.ylabel()` to set the title, x-axis label, and y-axis label, respectively. 

You can use `plt.xticks()` to rotate labels on the x-axis of a plot. The parameters `rotation=45, horizontalalignment='right'` can be passed in to rotate the labels by 45 degrees and align the labels to the right. 

You can use `plt.show()` to display a plot.

</details>

**Question:** What do you observe from this bar plot?

[Write your response here. Double-click (or enter) to edit.]

### Visualize the maximum unicorn company valuation per industry

Visualize unicorn companies' maximum valuation for each industry represented in the sample. To create a bar plot to visualize this, use the `bar()` function from the `matplotlib.pyplot` module. Before plotting, create a new column that represents the companies' valuations as numbers (instead of strings, as they're currently represented). Then, use this new column to plot your data.

In [ ]:
# Create a column representing company valuation as numeric data

In [ ]:
# Create bar plot
# with the various industries as the categories of the bars
# and the maximum valuation for each industry as the height of the bars

### YOUR CODE HERE ###


# Set title

### YOUR CODE HERE ###


# Set x-axis label

### YOUR CODE HERE ###


# Set y-axis label

### YOUR CODE HERE ###


# Rotate labels on the x-axis as a way to avoid overlap in the positions of the text  

### YOUR CODE HERE ###


# Display the plot

### YOUR CODE HERE ###


<details>
  <summary><h4><strong>Hint 1</strong></h4></summary>

Repeat the process from the last task, only this time with different variables.
</details>

<details>
  <summary><h4><strong>Hint 2</strong></h4></summary>

Use the function in the `matplotlib.pyplot` module that allows you to create a bar plot, specifying the category and height for each bar. 

Use the functions in the `matplotlib.pyplot` module that allow you to set the title, x-axis label, and y-axis label of plots. In that module, there are also functions for rotating the labels on the x-axis and displaying the plot. 

</details>

<details>
  <summary><h4><strong>Hint 3</strong></h4></summary>

Use the `plt.bar()` to create the bar plot, passing in the categories and heights of the bars.

Use `plt.title()`, `plt.xlabel()`, and `plt.ylabel()` to set the title, x-axis label, and y-axis label, respectively. 

Use `plt.xticks()` to rotate labels on the x-axis of a plot. The parameters `rotation=45, horizontalalignment='right'` can be passed in to rotate the labels by 45 degrees and align the labels to the right. 

Use `plt.show()` to display a plot.

</details>

**Question:** What do you observe from this bar plot? 

[Write your response here. Double-click (or enter) to edit.]

## Considerations

**What are some key takeaways that you learned from this lab?**

[Write your response here. Double-click (or enter) to edit.]

**What findings would you share with others?**

[Write your response here. Double-click (or enter) to edit.]

**What recommendations would you share with stakeholders based on these findings?**

[Write your response here. Double-click (or enter) to edit.]

**References**

Bhat, M.A. (2022, March). [*Unicorn Companies*](https://www.kaggle.com/datasets/mysarahmadbhat/unicorn-companies). 



**Congratulations!** You've completed this lab. However, you may not notice a green check mark next to this item on Coursera's platform. Please continue your progress regardless of the check mark. Just click on the "save" icon at the top of this notebook to ensure your work has been logged.